# CARM_VIS TESS LC PREPROCESSING FOR `LSTperiod`

This notebook preprocesses the TESS LC present in the CARM_VIS dataset to generate the proper datafiles for `LSTperiod`.

**IMPORTANT NOTE:** `LSTperiod` is based on Lomb-Scargle periodograms (as opposed to Generalized Lomb-Scargle periodograms), which means that it does not work well with curves with an offset value (i.e. not centered around the 0 value). Hence, for each light curve we must substract the median value before saving it with the proper format. The use of the median value instead of the mean value is to prevent outliers (be it stellar flares or just measurement errors) to greatly affect that mean, and hence the centering of the resulting curve.

## Modules and configuration

### Modules

In [10]:
import pandas as pd
import numpy as np

import lightkurve as lk

### Configuration

In [11]:
# FILES AND FOLDERS:
GTO_FILE = "../data/SELECTION_for_PG_CARM_VIS_objects_with_PG.csv"
TARGET_FOLDER = "./CARM_VIS_TESS_LC_LSTperiod/"


### Functions

In [12]:
def tess_lc_load(filename: str):
    '''Load the TESS LC file and returns a lightcurve object with just the three relevant columns
    Note: it seems to be necessary to drop 'nan' values for GLS to work properly.'''
    lc_lk = lk.read(filename).remove_nans()
    return lc_lk

## CARM_VIS objects file load

In [13]:
gto_df = pd.read_csv(GTO_FILE, sep=',', decimal='.')
gto_df.head()

Karmn             Name Comp    GJ  RA_J2016_deg  DE_J2016_deg  \
0  J23548+385  RX J2354.8+3831    -   NaN    358.713658     38.526340   
1  J23505-095       LP 763-012    -  4367    357.634705     -9.560964   
2  J23431+365          GJ 1289    -  1289    355.781509     36.536310   
3  J23381-162        G 273-093    -  4352    354.532687    -16.236514   
4  J23245+578       BD+57 2735    -   895    351.126628     57.853057   

      RA_J2000     DE_J2000  l_J2016_deg  b_J2016_deg  ... WF_offset_PG_TESS  \
0  23:54:51.46  +38:31:36.2   110.941908   -23.024449  ...        999.999756   
1  23:50:31.64  -09:33:32.7    80.777067   -67.303426  ...       1000.000122   
2  23:43:06.31  +36:32:13.1   107.922839   -24.336479  ...        999.999512   
3  23:38:08.16  -16:14:10.2    61.845437   -69.825220  ...       1000.000122   
4  23:24:30.51  +57:51:15.5   111.552287    -3.085183  ...        999.999512   

  WF_e_offset_PG_TESS  WF_FAP_PG_TESS WF_valid_PG_TESS  WF_error_PG_TESS  \
0        2.151008e-06             1.0              1.0               NaN   
1        9.022946e-07             1.0              1.0               NaN   
2        4.306074e-06             1.0              1.0               NaN   
3        9.022946e-07             1.0              1.0               NaN   
4        3.720858e-06             1.0              1.0               NaN   

   WF_elapsed_time_PG_TESS                                 WF_plain_file_TESS  \
0                94.758838  ../data/CARM_VIS_TESS_WinFunc_PGs/WF_J23548+38...   
1               132.607176  ../data/CARM_VIS_TESS_WinFunc_PGs/WF_J23505-09...   
2                97.939914  ../data/CARM_VIS_TESS_WinFunc_PGs/WF_J23431+36...   
3               136.603404  ../data/CARM_VIS_TESS_WinFunc_PGs/WF_J23381-16...   
4               131.327304  ../data/CARM_VIS_TESS_WinFunc_PGs/WF_J23245+57...   

                                    WF_fig_file_TESS  \
0  ../data/CARM_VIS_TESS_WinFunc_PGs/figures/WF_J...   
1  ../data/CARM_VIS_TESS_WinFunc_PGs/figures/WF_J...   
2  ../data/CARM_VIS_TESS_WinFunc_PGs/figures/WF_J...   
3  ../data/CARM_VIS_TESS_WinFunc_PGs/figures/WF_J...   
4  ../data/CARM_VIS_TESS_WinFunc_PGs/figures/WF_J...   

                                      PG_file_RV  \
0  ../data/CARM_VIS_RVs_PGs/J23548+385_RV_PG.dat   
1  ../data/CARM_VIS_RVs_PGs/J23505-095_RV_PG.dat   
2  ../data/CARM_VIS_RVs_PGs/J23431+365_RV_PG.dat   
3  ../data/CARM_VIS_RVs_PGs/J23381-162_RV_PG.dat   
4  ../data/CARM_VIS_RVs_PGs/J23245+578_RV_PG.dat   

                                     PG_file_TESS  
0  ../data/CARM_VIS_TESS_PGs/J23548+385_RV_PG.dat  
1  ../data/CARM_VIS_TESS_PGs/J23505-095_RV_PG.dat  
2  ../data/CARM_VIS_TESS_PGs/J23431+365_RV_PG.dat  
3  ../data/CARM_VIS_TESS_PGs/J23381-162_RV_PG.dat  
4  ../data/CARM_VIS_TESS_PGs/J23245+578_RV_PG.dat  

[5 rows x 300 columns]

In [14]:
len(gto_df)

269

In [15]:
print(list(gto_df.columns))

['Karmn', 'Name', 'Comp', 'GJ', 'RA_J2016_deg', 'DE_J2016_deg', 'RA_J2000', 'DE_J2000', 'l_J2016_deg', 'b_J2016_deg', 'Ref01', 'SpT', 'SpTnum', 'Ref02', 'Teff_K', 'eTeff_K', 'logg', 'elogg', '[Fe/H]', 'e[Fe/H]', 'Ref03', 'L_Lsol', 'eL_Lsol', 'Ref04', 'R_Rsol', 'eR_Rsol', 'Ref05', 'M_Msol', 'eM_Msol', 'Ref06', 'muRA_masa-1', 'emuRA_masa-1', 'muDE_masa-1', 'emuDE_masa-1', 'Ref07', 'pi_mas', 'epi_mas', 'Ref08', 'd_pc', 'ed_pc', 'Ref09', 'Vr_kms-1', 'eVr_kms-1', 'Ref10', 'ruwe', 'Ref11', 'U_kms-1', 'eU_kms-1', 'V_kms-1', 'eV_kms-1', 'W_kms-1', 'eW_kms-1', 'Ref12', 'sa_m/s/a', 'esa_m/s/a', 'Ref13', 'SKG', 'Ref14', 'SKG_lit', 'Ref14_lit', 'Pop', 'Ref15', 'vsini_flag', 'vsini_kms-1', 'evsini_kms-1', 'Ref16', 'P_d', 'eP_d', 'Ref17', 'pEWHalpha_A', 'epEWHalpha_A', 'Ref18', 'log(LHalpha/Lbol)', 'elog(LHalpha/Lbol)', 'Ref19', '1RXS', 'CRT_s-1', 'eCRT_s-1', 'HR1', 'eHR1', 'HR2', 'eHR2', 'Flux_X_E-13_ergcm-2s-1', 'eFlux_X_E-13_ergcm-2s-1', 'LX/LJ', 'eLX/LJ', 'Ref20', 'Activity', 'Ref21', 'FUV_mag',

## Create and store the relevant files for `LSTperiod`

Notice that we use the original TESS light curves to generate the files for LSTperiod, not the data stored in the periodogram calculation result file.

In [16]:
for i in range(0, len(gto_df)):
    lc_filename = "TESS-" + gto_df.loc[i, 'Karmn'] + ".dat" # Target filename for LSTperiod
    # Read the LC file:
    lc_lk = tess_lc_load(gto_df.loc[i, 'lc_file'])
    # Generate the data:
    lc_df = pd.DataFrame({'time': lc_lk['time'].value, 'value': lc_lk['flux'].value})
    # Remove the offset from the flux:
    median_flux = np.nanmedian(lc_df['value'])
    lc_df['value'] = lc_df['value'].map(lambda x: x - median_flux)
    # Save the file:
    lc_df.to_csv(TARGET_FOLDER + lc_filename, sep=' ', decimal='.', header=False, index=False)
    

# Summary

**OBSERVATIONS AND CONCLUSIONS:**
- We generated the 269 files of the TESS light curves in the proper format for `LSTperiod`.